### PR TITLE
[Sui]: Add SUI Implementation

### LINKED ISSUE #2882: [SUI]: Add SUI Network support
**Motivation**

[SUI](https://sui.io/): Sui is a boundless platform to build rich and dynamic on-chain assets from gaming to finance.

**Checklist**

- [x] Skeleton, registry.json update
- [x] Address support/Account 
- [x] Transaction support
    - [x] Transfer support
    - [x] Token transfer support
    - [x] NFT
    - [x] SC call
    - [ ] Staking

**Resources**

Slip id: 784
Derivation path: `Derivation path: m/44'/784'/0'/0'/0'`
[Source Code](https://github.com/MystenLabs/sui)
[Dev Doc](https://docs.sui.io/)
[TrustWallet Doc Blockchain Implementation](https://developer.trustwallet.com/wallet-core/newblockchain)

### PR DESCRIPTION
**Pull request by Milerius**
## Description

Fix #2882 

Todo:

- [x] Address
- [x] Ut address C++
- [x] Ut address KT
- [x] Ut address Swift
- [x] Transfer
- [x] Ut transfer
- [x] NFT transfer
- [x] Smart contract call

## How to test

Run C++ unit tests

## Checklist

- [x] Create pull request as draft initially, unless its complete.
- [x] Add tests to cover changes as needed.
- [x] Update documentation as needed.
- [x] If there is a related Issue, mention it in the description.

If you're adding a new blockchain

- [x] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain.

### CODE REVIEW

File: src/Move/Address.h [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+#pragma once
+
+#include "Data.h"
+#include "HexCoding.h"
+#include "PublicKey.h"
+#include <array>
+
+namespace TW::Move {
+template <typename Derived, std::size_t N>
+class Address {
```
Comment (hewigovens) [inline src/Move/Address.h:16]: I don't have a better name but `Move::Address` sounds a little bit weird, other Move chain might not have same address type

File: registry.json [hunk 2]
```diff
@@ -429,6 +429,35 @@
       "documentation": "https://fullnode.mainnet.aptoslabs.com/v1/spec#/"
     }
   },
+  {
+    "id": "sui",
+    "name": "Sui",
+    "displayName": "Sui",
```
Comment (hewigovens) [inline registry.json:435]:
```suggestion
```

File: src/Move/Address.h (hunk 1 above)
Comment (Milerius) [reply src/Move/Address.h:16]: It's actually named AccountAddress in Rust: https://github.com/move-language/move/blob/61b3aa7c7625968618984e783d820d90115ff694/language/move-core/types/src/account_address.rs#L14 - All Move chains should have the same address type, but could use different feature, aptos use `feature32` and sui `feature20`

File: android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/sui/TestSuiAddress.kt [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+    init {
+        System.loadLibrary("TrustWalletCore")
+    }
+
+    @Test
+    fun testAddress() {
+        val any = AnyAddress("0x061ce2b2100a71bb7aa0da98998887ad82597948", CoinType.SUI)
+        assertEquals(any.coin(), CoinType.SUI)
+        assertEquals(any.description(), "0x061ce2b2100a71bb7aa0da98998887ad82597948")
```
Comment (hewigovens) [inline android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/sui/TestSuiAddress.kt:26]: It seems this is a valid Ethereum address too, user will be easy to make mistakes, can we ask Sui team that if they have ever considered this issue?

File: rust/src/encoding/mod.rs [hunk 4]
```diff
@@ -0,0 +1,63 @@
+use base64::{Engine as _, engine::{general_purpose}};
```
Comment (hewigovens) [inline rust/src/encoding/mod.rs:1]: license header

File: src/Base64.cpp [hunk 5]
```diff
@@ -97,4 +102,21 @@ string encodeBase64Url(const Data& val) {
     return encoded;
 }
 
+std::string encode_rust(const Data& val) {
```
Comment (hewigovens) [inline src/Base64.cpp:105]: what's the difference of new added base64 encoding?

File: src/proto/Sui.proto [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+syntax = "proto3";
+
+package TW.Sui.Proto;
+option java_package = "wallet.core.jni.proto";
+
+message SigningInput {
+    bytes private_key = 1;
+    // Obtain by calling any write RpcJson on SUI
+    string unsigned_tx = 2;
```
Comment (hewigovens) [inline src/proto/Sui.proto:15]: this `unsigned_tx` is barely unreadable, at least add simple messages like `transfer`, `contract call`

File: src/Base64.cpp (hunk 5 above)
Comment (Milerius) [reply src/Base64.cpp:105]: It's seems that boost base64 iterator is not compliant, (see the unit tests for decode sui in Base64Tests.cpp - this one is compliant.

File: android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/sui/TestSuiAddress.kt (hunk 3 above)
Comment (Milerius) [reply android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/sui/TestSuiAddress.kt:26]: I forwarded the question

File: src/proto/Sui.proto [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    string unsigned_tx_msg = 1;
+}
+
+// Input data necessary to create a signed transaction.
+message SigningInput {
+    //  Private key to sign the transaction (bytes)
+    bytes private_key = 1;
+
+    oneof transaction_payload {
+        SignDirect sign_direct_message = 2;
```
Comment (hewigovens) [inline src/proto/Sui.proto:24]: Is there a way to sign simple transfer without calling rpc?

File: src/Base64.cpp (hunk 5 above)
Comment (hewigovens) [reply src/Base64.cpp:105]: in this case, maybe worth to remove using boost base64

File: src/proto/Sui.proto (hunk 7 above)
Comment (Milerius) [reply src/proto/Sui.proto:24]: answered in DM

File: include/TrustWalletCore/TWBlockchain.h [hunk 8]
```diff
@@ -57,6 +57,7 @@ enum TWBlockchain {
     TWBlockchainAptos = 43, // Aptos
     TWBlockchainHedera = 44, // Hedera
     TWBlockchainTheOpenNetwork = 45,
+    TWBlockchainSui = 46,  // TODO remove if the blockchain already exists, or just remove this comment if not
```
Comment (catenocrypt) [inline include/TrustWalletCore/TWBlockchain.h:60]: Remove TODO comment
```suggestion
    TWBlockchainSui = 46,
```

**Review by catenocrypt — APPROVED** [review]

**Review by hewigovens — DISMISSED** [review]

**Review by hewigovens — APPROVED** [review]

**Review by catenocrypt — APPROVED** [review]

**Pull request merged**